Feature/modify hca human experiments endpoint

upendrakumbham opened the pull request:
The goal of this feature is: Modify HCA Human experiments endpoint("/json/experiments/hca/human") to accept multiple organism parts. 

Example:` http://localhost:8080/sc/json/experiments/hca/human?organismPart=lymph node&organismPart=skin`

Changed Solr query to accept multiple organism parts.

In this PR we have switched from normal query to streaming expressions to improve performance. Here are the details for the proof of performance boost.
Pivotal story references - https://www.pivotaltracker.com/story/show/172792515
                                            https://www.pivotaltracker.com/story/show/172101611

upendrakumbham committed:
Modify RequestParam from String to Set

upendrakumbham committed:
Modify method

-Added Set as a method parameter

upendrakumbham committed:
Modify Solr query to accept multiple Organism parts

- Used streaming expressions to optimise Solr query.

upendrakumbham committed:
Fix broken tests

upendrakumbham committed:
Change SolrQuery to search based on one of the values of ontology annotation,ontology annotation label and its ancestors

upendrakumbham committed:
Remove extra space

upendrakumbham committed:
Fix broken tests

upendrakumbham committed:
Add missing scenarios

upendrakumbham committed:
Remove default value

upendrakumbham committed:
Rename tests

upendrakumbham committed:
Change Assertion statements

- I have a feeling of that this test would fail in the production if we have more human experiments in the Postgres database if we compare with size as we are relying on the actual database instead of test database. I'm might be wrong.

upendrakumbham committed:
Code alignment

upendrakumbham committed:
Remove extra space

upendrakumbham committed:
Remove value attribute as it is not required and it makes to fail empty parameters use case.

- If we pass just ex: http://localhost:8080/sc/json/experiments/hca/human -
it will look for the value 'organismPart' for the RequestParam to call endpoint.
  It would make web integration tests to fail.

upendrakumbham committed:
Moved all dao mocking to the relevant test case

- Reason for moving is: When we call @Setup, for each testcase all the dao mock scenarios were execute which we don't need for each testcase.

upendrakumbham committed:
Add testcase for empty characteristic name and value

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -18,41 +22,46 @@
 
-    public ImmutableSet<String> fetchExperimentAccessions(String characteristicName, String characteristicValue) {
+    public ImmutableSet<String> fetchExperimentAccessions(String characteristicName,
+                                                          Set<String> characteristicValues) {
```
it would be preferable to use `ImmutableSet<String>` instead of `Set<String>`

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -18,41 +22,46 @@
-                .setNormalize(false)
-                .addQueryFieldByTerm(CHARACTERISTIC_NAME, characteristicName)
+
+        queryBuilder.addQueryFieldByTerm(CHARACTERISTIC_NAME, characteristicName)
```
We need the normalize flag to exist for the cases when we just pass ontology id `e.g. UBERON_0000029`

When the normalize flag is not set then the resulting solr query is not correctly formed, which in this case would be `q=characteristic_name:("organism_part") AND characteristic_value("UBERON_0000029")...` which would return empty result set as there is no field with `characteristic_value: UBERON_0000029`.

So with the normalizing flag set to `false` the resulting query would be `q=characteristic_name:("organism_part") AND characteristic_value:*UBERON_0000029 ....` which is what we need.

upendrakumbham replied:
Yes, it make sense. I fixed it.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -18,41 +22,46 @@
-                            ONTOLOGY_ANNOTATION_LABEL, ImmutableList.of("*".concat(characteristicValue))
-                    ));
+                .sortBy(EXPERIMENT_ACCESSION, SolrQuery.ORDER.asc)
+                .setRows(SolrQueryBuilder.DEFAULT_ROWS);
```
The default number of rows would not give us the correct result in case of this query because the result set is quite big. For example, when I ran test case `returnAllExperimentsForEmptyCharacteristicValue` with default rows I got the result set of just `14 experiment accessions` whereas when I ran with previous row numbers `10000000` I got a result set of `150 experiment accessions` which is the correct number of experiment accessions we expect to get. This is the result when I ran it on our wwwdev solr instance.

upendrakumbham replied:
Fixed it.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -18,41 +22,46 @@
+                    ONTOLOGY_ANNOTATION_ANCESTORS_LABELS, characteristicValues,
+                    ONTOLOGY_ANNOTATION_ANCESTORS_URIS, characteristicValues,
+                    ONTOLOGY_ANNOTATION_LABEL, characteristicValues
+            ));
```
as mentioned above we would need the concatenation of `*` in the case when just `ontology_id` is passed

upendrakumbham replied:
Fixed it.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -18,41 +22,46 @@
+            return (tupleStreamer.get()
+                    .map(tuple -> (String) tuple.get(EXPERIMENT_ACCESSION.name()))
+                    .collect(toImmutableSet()));
+        }
```
Why did we go from normal solr query to using solr stream expressions?

I believe in this case the performance with simple solr query is much faster than using stream expressions. For example when I ran test case `returnAllExperimentsForEmptyCharacteristicValue` with stream expression my laptop took `59 secs` to finish the test case whereas with normal solr query it just took `19 secs`.

upendrakumbham replied:
Why did we go from normal solr query to using solr stream expressions? - the approach suggested by Alfonso. We will check with him what he would suggest before changing.

haideriqbal replied:
Regarding this, I guess once you have profiled what is the bottleneck in solr java code then we can decide which approach to take.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:
```diff
@@ -18,11 +20,16 @@ public HcaHumanExperimentService(ExperimentTrader experimentTrader,
-    public ImmutableSet<Experiment> getPublicHumanExperiments(String characteristicName, String characteristicValue) {
-        return hcaHumanExperimentDao.fetchExperimentAccessions(characteristicName, characteristicValue)
+    public ImmutableSet<Experiment> getPublicHumanExperiments(String characteristicName,
+                                                              Set<String> characteristicValues) {
```
Same suggestion as above. Make more use of `ImmutableSet<String>` instead of `Set<String>`

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:
```diff
@@ -18,11 +20,16 @@ public HcaHumanExperimentService(ExperimentTrader experimentTrader,
                 .filter(experiment -> experiment.getSpecies().isUs())
                 .collect(toImmutableSet());
     }
+
```
Remove blank space.

upendrakumbham replied:
The tool is giving this extra black space before and end of class. 
I am not sure how the tool is working in you're development IDE.
Please, let me know to fix it if required.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:
```diff
@@ -18,11 +20,16 @@ public HcaHumanExperimentService(ExperimentTrader experimentTrader,
+    public ImmutableSet<Experiment> getPublicHumanExperiments(String characteristicName,
+                                                              Set<String> characteristicValues) {
+
+        Set<String> experimentAccessions = hcaHumanExperimentDao.fetchExperimentAccessions(characteristicName,
```
`experimentAccessions` is a throw away variable here. You can directly do 
`return hcaHumanExperimentDao.fetchExperimentAccessions(characteristicName, characteristicValues)
.stream()
....`

upendrakumbham replied:
Removed throw away variable.

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java:
```diff
@@ -21,7 +26,9 @@ public HcaHumanExperimentsController(HcaHumanExperimentService hcaHumanExperimen
             produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
-    public String getOrganismPartExperimentsList(@RequestParam(defaultValue = "") String organismPart) {
+    public String getOrganismPartExperimentsList(@RequestParam(defaultValue = "") Set<String> organismPart) {
+        LOGGER.info("Number of organism parts: {}",organismPart.size());
```
I guess there is no need for logging this information

upendrakumbham replied:
Removed.

haideriqbal commented on src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:
```diff
@@ -33,27 +34,51 @@ void setUp() {
     void notEmptyForWhenValidOnotologyIdIsPassed() {
-        assertThat(subject.fetchExperimentAccessions("organism_part", "UBERON_0000014"))
+        assertThat(subject.fetchExperimentAccessions("organism_part",
+                ImmutableSet.of("http://purl.obolibrary.org/obo/UBERON_0000029")))
```
We need to pass just `ontologyId` as this is what we will be getting from `Anatomogram` instead of complete url

upendrakumbham replied:
Got it. Fixed.

haideriqbal commented on src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:
```diff
@@ -33,27 +34,51 @@ void setUp() {
+        assertThat(subject.fetchExperimentAccessions("organism_part", ImmutableSet.of()))
                 .isNotEmpty();
     }
+
+    @Test
+    void notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheUrlIsPassed() {
+        assertThat(subject.fetchExperimentAccessions("organism_part",
+                ImmutableSet.of("http://www.ebi.ac.uk/efo/EFO_0000001")))
+                .isNotEmpty()
+                .contains("E-EHCA-2");
+    }
+
+    @Test
+    void notEmptyWhenValidOntologyAnnotationLabelIsPassed() {
+        assertThat(subject.fetchExperimentAccessions("organism_part", ImmutableSet.of("zone of skin")))
+                .isNotEmpty()
+                .contains("E-EHCA-2");
+    }
+
+    @Test
+    void notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheLabelIsPassed() {
+        assertThat(subject.fetchExperimentAccessions("organism_part", ImmutableSet.of("anatomical structure")))
+                .isNotEmpty()
+                .contains("E-EHCA-2");
+    }
+
```
These tests are not necessary for our use case but good to be cautious 👍

haideriqbal commented on src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:
```diff
@@ -33,27 +34,51 @@ void setUp() {
 
```
Please also add a test case which has multiple `organism_parts` in the set to check if the functionality is working correctly or not.

upendrakumbham replied:
Added test case.

haideriqbal commented on src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:
```diff
@@ -27,37 +27,50 @@
+    }
+
+    @Test
+    public void ShouldGetAllHumanExperimentsWithEmptyCharacteristicValue() {
```
very minor but the name of the method starts with a lower case letter 🕵️‍♂️

upendrakumbham replied:
Sorry. Silly mistake.

haideriqbal commented on src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:
```diff
@@ -27,37 +27,50 @@
-                subject.getPublicHumanExperiments("organism_part","foo"))
-                .hasSize(0);
+    public void ShouldGetAllHumanExperimentsWithEmptyCharacteristicNameAndValue() {
+        when(hcaHumanExperimentDaoMock.fetchExperimentAccessions("", ImmutableSet.of()))
```
If you move `when(...) then(...)` statements in the `setUp` method then you would not need to repeat this statement in multiple methods also the test case themselves would be much cleaner and simple.

upendrakumbham replied:
It's true if we are using the same mocking in multiple test cases. 
But we are not repeating mock scenarios, they are different from each other.
If we move them back to setUp(), all mock scenarios would execute for every test case. 
Let me know if you are not happy with this, I can move them back to setUP() as you're the owner of this code :).

haideriqbal replied:
Ohh I missed it earlier you are right they are different for each test so makes sense 👍 
Regarding ownership of the code, we all are owners of it as a team 😄

haideriqbal reviewed (changes requested):
Please look at the comments. There are few changes which are required otherwise good work 👍

upendrakumbham commented:
Thanks for the review comments, Haider. I will have a look.

upendrakumbham committed:
Add normalize flag(false)

- When the normalize flag is not set then the resulting solr query is not correctly formed, which in this case would be q=characteristic_name:("organism_part") AND characteristic_value("UBERON_0000029")... which would return empty result set as there is no field with characteristic_value: UBERON_0000029.

So with the normalizing flag set to false the resulting query would be q=characteristic_name:("organism_part") AND characteristic_value:*UBERON_0000029 .... which is what we need.

upendrakumbham committed:
Update solr query rows to '10000000'

- The default number of rows would not give us the correct result in case of this query because the result set is quite big. For example, when I ran test case returnAllExperimentsForEmptyCharacteristicValue with default rows I got the result set of just 14 experiment accessions whereas when I ran with previous row numbers 10000000 I got a result set of 150 experiment accessions which is the correct number of experiment accessions we expect to get. This is the result when I ran it on our wwwdev solr instance.

upendrakumbham committed:
Add method to modify collection(*)

- Concat all the elements with * in the given collection and return modified collection

upendrakumbham committed:
Update Collection using concatStar()

upendrakumbham committed:
Remove logger

upendrakumbham committed:
code reformat

upendrakumbham committed:
Add only OnotologyId

upendrakumbham committed:
Add test case for multiple characteristic values

upendrakumbham committed:
Follow camel casing

upendrakumbham committed:
Remove throw away variable

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:
```diff
@@ -64,4 +65,11 @@ public HcaHumanExperimentDao(SolrCloudCollectionProxyFactory solrCloudCollection
         }
     }
+
+    private ImmutableSet<String> concatStar(Set<String> characteristicValues) {
```
🏅

haideriqbal commented on src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java:
```diff
@@ -27,13 +27,12 @@ public HcaHumanExperimentsController(HcaHumanExperimentService hcaHumanExperimen
                         .map(ExperimentJsonSerializer::serialize)
                         .collect(toImmutableSet()));
     }
+
```
remove blank space, please

haideriqbal reviewed (changes requested):
Good work! Let's see what comes out of solr profiling and then we can decide and merge this PR.

haideriqbal commented:
Performance appears to have improved significantly with new sharding in the solr. Just as a last curious step, can we run the code with streaming expressions and see how it has improved? The last time we ran we were getting around ~50 sec time.

upendrakumbham commented:
Hi @haideriqbal, I am trying to run a few integration tests on this branch to proof code performance, But I am getting compilation issues on this branch because of this class `'ExperimentJsonSerializer'`. This has been modified by you in the core project after this PR. I took the latest code from the core and rebuild it. But still, I am getting the same error. Could you please try to run those tests if you can. It would be very helpful.
-Thank you

alfonsomunozpomer commented:
@upendrakumbham Can you please check out a “compatible” commit in `atlas-web-core`? I think the commit right before 8aacc2c9f250891821d4bb468ffe85f02b35e6e2 in the master branch will work. Please, give that a go and let us know.

upendrakumbham commented:
`git checkout 6ddb790730e0126967d2d5d84c33d014c4401678` core master branch. All errors gone.

Below tests, I ran:
```
uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT:

3.82 spassed notEmptyForWhenValidOnotologyIdIsPassed()
1 m 10 s passed notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheLabelIsPassed()
3.96 s passed notEmptyForCorrectCharacteristicType()
1 m 28 spassedreturnAllExperimentsForEmptyCharacteristicValue()
3.74 spassednotEmptyWhenValidOntologyAnnotationLabelIsPassed()
122 ms passed emptyIfNoExperimentsCanBeFound()
4.76 s passed returnMatchingExperimentsForMultipleCharacteristicValues()
1 m 20 s passed notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheUrlIsPassed()

```

upendrakumbham commented:
git checkout  e9cece92985aecfee59967f189f43eddcf34d098. Core master branch.

I got `setFactes` related compilation errors in the `genesearchdao` and `cellmedatadatadao.  As they are very few I just commented out and ran tests.

`uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT:
-----
2.44 s passednotEmptyForWhenValidOnotologyIdIsPassed()
1 m 27 s passednotEmptyWhenValidOntologyAnnotationAncestorsOneOfTheLabelIsPassed()
5.22 s passednotEmptyForCorrectCharacteristicType()
1 m 37 s passedreturnAllExperimentsForEmptyCharacteristicValue()
4.09 s passednotEmptyWhenValidOntologyAnnotationLabelIsPassed()
126 ms passedemptyIfNoExperimentsCanBeFound()
5.41 s passedreturnMatchingExperimentsForMultipleCharacteristicValues()
1 m 29 s passednotEmptyWhenValidOntologyAnnotationAncestorsOneOfTheUrlIsPassed()

upendrakumbham commented:
Above one right before 8aacc2c9f250891821d4bb468ffe85f02b35e6e2 in the master branch.

upendrakumbham commented:
@Haider- you can update your results if you have any.

haideriqbal commented:
I am getting this test `notEmptyForCorrectCharacteristicType` failed. Can you please check!

Regarding timings stat, I will leave it up to you to judge as the stats on my system are dodgy atm. If you think this is faster now let's go with streaming expressions then.

upendrakumbham commented:
Yes, we are getting more than one experiment_acession for the characteristic_value 'skin'. That might be the reason, at least for me the reason for failed tests. Modified assertion to fix this.

Ran tests using Solr normal query:
--
Here are the results:

```
 uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT
--
5.02 s passed notEmptyForWhenValidOnotologyIdIsPassed()
 6.11 s passed notEmptyForCorrectCharacteristicType()
 2 m 29 s passed returnAllExperimentsForEmptyCharacteristicValue()
117 ms passed. emptyIfNoExperimentsCanBeFound()
```

haideriqbal reviewed (approved):
Cool! Can you please fix the assertions and commit it. Then you can resolve the conflicts and merge the PR.

upendrakumbham commented:
` 2 m 29 s passed returnAllExperimentsForEmptyCharacteristicValue()` this is clearly saying that compare to normal Solr query streaming expressions reduced half of the time. So It proved to us that streaming expression boosting performance. So we are going with streaming expressions finally.

upendrakumbham commented:
> Cool! Can you please fix the assertions and commit it. Then you can resolve the conflicts and merge the PR.

Yes, I will.

upendrakumbham committed:
Remove unwanted test cases

 - These test scenarios are not really required  and Solr paying bit more time to run these.

upendrakumbham committed:
Revert "Remove unwanted test cases"

This reverts commit 4609a7dd05d00a73bc903e791a0792c3a7dcec72.

upendrakumbham committed:
Remove a few test scenarios

Removed a few scenarios which are not that worth  and Solr also paying more time for them to run.

upendrakumbham committed:
Fix assertions

 some times for one characteristic_value 'skin', we are getting more than one experiment_accessions.

upendrakumbham commented:
This PR got merged with the local master and resolved all the conflicts manually. I have tested the code in the local and pushed it to the remote.
Verified Bamboo, got a green build. Finally, this story is closed :).

upendrakumbham commented:
this PR has been merged to master trough the command line. So I have closed this PR.
